File: frontend/src/lib/api/purchases.ts
import { apiClient } from './client.ts';
import type { Purchase } from '../../types/index.ts';

export async function getPurchases(params?: { supplier_id?: string; status?: string }): Promise<Purchase[]> {
  return apiClient.get<Purchase[]>('/purchases', { params });
}

export interface CreatePurchaseRequest {
  supplier_id: string;
  items: Array<{
    variant_id: string;
    quantity: number;
    unit_cost: number;
  }>;
  invoice_number?: string;
  discount?: number;
  paid_amount?: number;
}

export async function createPurchase(data: CreatePurchaseRequest): Promise<Purchase> {
  return apiClient.post<Purchase>('/purchases', data);
}

export async function getPurchase(purchaseId: string): Promise<Purchase> {
  return apiClient.get<Purchase>(`/purchases/${purchaseId}`);
}

export interface UpdatePurchaseRequest {
  supplier_id?: string;
  invoice_number?: string | null;
  discount?: number;
  items?: Array<{
    variant_id: string;
    quantity: number;
    unit_cost: number;
  }>;
}

export async function updatePurchase(purchaseId: string, data: UpdatePurchaseRequest): Promise<Purchase> {
  return apiClient.patch<Purchase>(`/purchases/${purchaseId}`, data);
}

export async function deletePurchase(purchaseId: string): Promise<void> {
  await apiClient.delete(`/purchases/${purchaseId}`);
}

File: backend/app/schemas/purchases.py
"""Response schemas for the purchases domain (Step 3).

These mirror the SQLAlchemy models in the purchases service.
"""

from datetime import datetime
from uuid import UUID
from decimal import Decimal

from pydantic import BaseModel, ConfigDict

from app.models.payment import PaymentMethod


class PurchaseItemResponse(BaseModel):
    """One line on a purchase."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    variant_id: UUID
    quantity: Decimal
    unit_cost: Decimal
    total: Decimal


class PurchaseResponse(BaseModel):
    """A full purchase with its lines."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    supplier_id: UUID
    invoice_number: str | None = None
    subtotal: Decimal
    discount: Decimal
    total: Decimal
    paid_amount: Decimal
    due_amount: Decimal
    shop_id: UUID
    created_at: datetime
    items: list[PurchaseItemResponse] = []


class PurchaseListItemResponse(BaseModel):
    """Compact purchase for list views."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    shop_id: UUID
    order_number: str | None = None
    supplier_id: UUID
    supplier_name: str | None = None
    status: str = "received"
    total_amount: Decimal
    paid_amount: Decimal
    items_count: int = 0
    created_at: datetime


class CreatePurchaseRequest(BaseModel):
    supplier_id: UUID
    items: list["PurchaseLineRequest"]
    invoice_number: str | None = None
    discount: Decimal = 0
    paid_amount: Decimal = 0


class PurchaseLineRequest(BaseModel):
    variant_id: UUID
    quantity: float
    unit_cost: float
